ti/instrumentation/instrumentation.go

```diff
@@ -132,23 +135,44 @@ func checkForBazelOptimization(ctx context.Context, workspace string, fs filesys
 // ComputeSelectedTestsV2 updates TI selection depending on the split strategy
 // AutoDetectTests output and parallelism configuration
 func ComputeSelectedTestsV2(ctx context.Context, runConfigV2 *api.RunTestsV2Config, log *logrus.Logger,
-	selection *ti.SelectTestsResp, stepID, workspace string, envs map[string]string, tiConfig *tiCfg.Cfg) {
+	selection *ti.SelectTestsResp, stepID, workspace string, envs map[string]string, tiConfig *tiCfg.Cfg, runOnlySelectedTests bool, fs filesystem.FileSystem) bool {
 	// Adding only this remove this condition later when we have complete specs
-	if len(selection.Tests) == 0 {
+	if runOnlySelectedTests && len(selection.Tests) == 0 {
 		// TI returned zero test cases to run. Skip parallelism as
 		// there are no tests to run
-		return
+		return runOnlySelectedTests
 	}
 	splitIdx, splitTotal := GetSplitIdxAndTotal(envs)
 	tests := make([]ti.RunnableTest, 0)
 
-	//TODO:V2 Autodetection
-
-	if len(selection.Tests) > 0 {
+	// Autodetection
+	if !runOnlySelectedTests {
+		// For full runs, detect all the tests in the repo and split them
+		// If autodetect fails or detects no tests, we run all tests in step 0
+		var err error
+		testGlobs := runConfigV2.TestGlobs
+		tests, err = AutoDetectTests(ctx, workspace, testGlobs, log, envs, fs)
+		if err != nil || len(tests) == 0 {
+			// AutoDetectTests output should be same across all the parallel steps. If one of the step
+			// receives error / no tests to run, all the other steps should have the same output
+			if splitIdx == 0 {
+				// Error while auto-detecting, run all tests for parallel step 0
+				runOnlySelectedTests = false
+				log.Errorln("Error in auto-detecting tests for splitting, running all tests")
+			} else {
+				// Error while auto-detecting, no tests for other parallel steps
+				selection.Tests = []ti.RunnableTest{}
+				runOnlySelectedTests = true
```

Comment:
why we are doing "runOnlySelectedTests = true" here

Reply:
so that, in case there is any error during test selection and its not 0th step during parallelism, it should not run anything, filter file will be populated with 0 cases.